modules/auxiliary/scanner/http/log4shell_scanner.rb

```diff
@@ -0,0 +1,198 @@
+      peerinfo = "#{target_info[:rhost]}:#{target_info[:rport]}"
+      print_good("#{peerinfo.ljust(21)} - Log4Shell found via #{details}")
+      report_vuln(
+        host: target_info[:rhost],
+        port: target_info[:rport],
+        info: "Module #{fullname} detected Log4Shell vulnerability via #{details}",
+        name: name,
+        refs: references
+      )
+    end
+  rescue Net::LDAP::PDU::Error => e
+    vprint_error("#{peer} - #{e}")
+  ensure
+    service.close_client(client)
+  end
+
+  def rand_text_alpha_lower_numeric(len, bad = '')
+    foo = []
+    foo += ('a'..'z').to_a
+    foo += ('0'..'9').to_a
+    Rex::Text.rand_base(len, bad, *foo)
+  end
+
+  def run
+    fail_with(Failure::BadConfig, 'The SRVHOST option must be set to a routable IP address.') if ['0.0.0.0', '::'].include?(datastore['SRVHOST'])
+    @mutex = Mutex.new
+    @tokens = {}
+    # always disable SSL because the LDAP server doesn't use it but the setting is shared with the HTTP requests
+    begin
+      start_service('SSL' => false)
+    rescue Rex::BindFailed => e
+      fail_with(Failure::BadConfig, e.to_s)
+    end
+
+    super
+
+    print_status("Sleeping #{datastore['LDAP_TIMEOUT']} seconds for any last LDAP connections")
+    sleep datastore['LDAP_TIMEOUT']
+  ensure
+    stop_service
+  end
+
+  def replicant
+    #
+    # WARNING: This is a horrible pattern and should not be copy-pasted into new code. A better solution is currently
+    # in the works to address service / socket replication as it affects scanner modules.
+    #
+    service = @service
+    @service = nil
+    obj = super
+    @service = service
+
+    # but do copy the tokens and mutex to the new object
+    obj.mutex = @mutex
+    obj.tokens = @tokens
+    obj
+  end
+
+  def run_host(ip)
+    # probe the target before continuing
+    return if send_request_cgi('uri' => normalize_uri(target_uri)).nil?
+
+    run_host_uri(ip, normalize_uri(target_uri)) unless target_uri.blank?
+
+    return if datastore['URIS_FILE'].blank?
+
+    File.open(datastore['URIS_FILE'], 'rb').each_line(chomp: true) do |uri|
+      next if uri.blank? || uri.start_with?('#')
+
+      if uri.include?('${jndi:uri}')
+        token = rand_text_alpha_lower_numeric(8..32)
+        jndi = jndi_string(token)
+        uri.delete_prefix!('/')
+        test(token, uri: normalize_uri(target_uri, '') + uri.gsub('${jndi:uri}', Rex::Text.uri_encode(jndi)))
+      else
+        run_host_uri(ip, normalize_uri(target_uri, uri))
+      end
+    end
+  end
+
+  def run_host_uri(_ip, uri)
+    unless datastore['HEADERS_FILE'].blank?
+      headers_file = File.open(datastore['HEADERS_FILE'], 'rb')
+      headers_file.each_line(chomp: true) do |header|
+        next if header.blank? || header.start_with?('#')
+
+        token = rand_text_alpha_lower_numeric(8..32)
+        test(token, uri: uri, headers: { header => jndi_string(token) })
+      end
+    end
+
+    token = rand_text_alpha_lower_numeric(8..32)
```

Comment:
Not a blocker: Looks like there's a small possibility of creating a duplicate token here, presumably more likely when two threads create a token of length 8 😄

Reply:
There's a `Rex::Text` method for this, why an internal impl?

Reply:
Are you sure there's a Rex::Text method for this? I didn't see one that uses this character set. I wanted it lower and numeric to ensure that it was case insensitive.